sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileScanRDD.scala

```diff
@@ -103,6 +116,108 @@ class FileScanRDD(
+        if (metadataColumns.nonEmpty) {
+          if (currentFile == null) {
+            metadataColumnsUnsafeRow = null
+            metadataColumnsInternalRow = null
+          } else {
+            // construct an internal row
+            val path = new Path(currentFile.filePath)
+            metadataColumnsInternalRow = InternalRow.fromSeq(
+              metadataColumns.map(_.name).map {
+                case FILE_PATH => UTF8String.fromString(path.toString)
+                case FILE_NAME => UTF8String.fromString(path.getName)
+                case FILE_SIZE => currentFile.fileSize
+                case FILE_MODIFICATION_TIME => currentFile.modificationTime
+              }
+            )
+            // convert the internal row to an unsafe row
+            metadataColumnsUnsafeRow = unsafeRowConverter(metadataColumnsInternalRow)
+          }
+        }
+      }
+
+      /**
+       * Create a writable column vector containing all required metadata columns
+       */
+      private def createMetadataColumnVector(c: ColumnarBatch): Array[WritableColumnVector] = {
+        val path = new Path(currentFile.filePath)
+        val filePathBytes = path.toString.getBytes
+        val fileNameBytes = path.getName.getBytes
+        var rowId = 0
+        metadataColumns.map(_.name).map {
+          case FILE_PATH =>
+            val columnVector = new OnHeapColumnVector(c.numRows(), StringType)
+            rowId = 0
+            // use a tight-loop for better performance
+            while (rowId < c.numRows()) {
+              columnVector.putByteArray(rowId, filePathBytes)
+              rowId += 1
+            }
+            columnVector
+          case FILE_NAME =>
+            val columnVector = new OnHeapColumnVector(c.numRows(), StringType)
+            rowId = 0
+            // use a tight-loop for better performance
+            while (rowId < c.numRows()) {
+              columnVector.putByteArray(rowId, fileNameBytes)
+              rowId += 1
+            }
+            columnVector
+          case FILE_SIZE =>
+            val columnVector = new OnHeapColumnVector(c.numRows(), LongType)
+            columnVector.putLongs(0, c.numRows(), currentFile.fileSize)
+            columnVector
+          case FILE_MODIFICATION_TIME =>
+            val columnVector = new OnHeapColumnVector(c.numRows(), LongType)
+            columnVector.putLongs(0, c.numRows(), currentFile.modificationTime)
+            columnVector
+        }.toArray
+      }
+
+      /**
+       * Add metadata columns at the end of nextElement if needed.
+       * For different row implementations, use different methods to update and append.
+       */
+      private def addMetadataColumnsIfNeeded(nextElement: Object): Object = {
+        if (metadataColumns.nonEmpty) {
+          nextElement match {
+            case c: ColumnarBatch =>
+              new ColumnarBatch(
+                Array.tabulate(c.numCols())(c.column) ++ createMetadataColumnVector(c),
```

Comment:
```suggestion
                c.column ++ createMetadataColumnVector(c),
```

Reply:
I think we have to do `Array.tabulate(c.numCols())(c.column)`? since we need to get `c.colmn(ordinal)` for all ordinals from `[0 - c.numCols)`.

Reply:
hmm, is `Array(c.column(0), c.column(1), ... c.column(numCols - 1))` the same as `c.column`?

Reply:
`c.column` needs to take an ordinal, or we could make `protected final ColumnVector[] columns;` -> `public`?

Reply:
ah sorry, I was confused by java code vs scala code. nvm, the code here is correct.